modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/StorageScanNode.java

```diff
@@ -0,0 +1,303 @@
+import org.apache.ignite.internal.sql.engine.exec.RowHandler;
+import org.apache.ignite.internal.sql.engine.schema.InternalIgniteTable;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Based abstract scan node.
+ */
+public abstract class StorageScanNode<RowT> extends AbstractNode<RowT> {
+    /** Special value to highlights that all row were received and we are not waiting any more. */
+    private static final int NOT_WAITING = -1;
+
+    private final Queue<RowT> inBuff = new LinkedBlockingQueue<>(inBufSize);
+
+    private final @Nullable Predicate<RowT> filters;
+
+    private final @Nullable Function<RowT, RowT> rowTransformer;
+
+    private final Function<BinaryRow, RowT> tableRowConverter;
+
+    private int requested;
+
+    private int waiting;
+
+    private boolean inLoop;
+
+    private Subscription activeSubscription;
+
+    /**
+     * Constructor.
+     *
+     * @param ctx Execution context.
+     * @param rowFactory Row factory.
+     * @param schemaTable The table this node should scan.
+     * @param filters Optional filter to filter out rows.
+     * @param rowTransformer Optional projection function.
+     */
+    public StorageScanNode(
+            ExecutionContext<RowT> ctx,
+            RowHandler.RowFactory<RowT> rowFactory,
+            InternalIgniteTable schemaTable,
+            @Nullable Predicate<RowT> filters,
+            @Nullable Function<RowT, RowT> rowTransformer,
+            @Nullable BitSet requiredColumns
+    ) {
+        super(ctx);
+
+        assert context().transaction() != null || context().transactionTime() != null : "Transaction not initialized.";
+
+        tableRowConverter = row -> schemaTable.toRow(context(), row, rowFactory, requiredColumns);
+
+        this.filters = filters;
+        this.rowTransformer = rowTransformer;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void request(int rowsCnt) throws Exception {
+        assert rowsCnt > 0 && requested == 0 : "rowsCnt=" + rowsCnt + ", requested=" + requested;
+
+        checkState();
+
+        requested = rowsCnt;
+
+        if (!inLoop) {
+            context().execute(this::push, this::onError);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void closeInternal() {
+        super.closeInternal();
+
+        if (activeSubscription != null) {
+            activeSubscription.cancel();
+
+            activeSubscription = null;
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    protected void rewindInternal() {
+        requested = 0;
+        waiting = 0;
+
+        if (activeSubscription != null) {
+            activeSubscription.cancel();
+
+            activeSubscription = null;
+        }
+    }
+
+    /**
+     *  Publisher of scan node.
+     *
+     *  @return Publisher of scan node or {@code null} in case nothing to scan.
+     */
+    protected abstract Publisher<RowT> scan();
```

Comment:
```suggestion
    protected abstract @Nullable Publisher<RowT> scan();
```

Contract of this method is not clear. The fact you need to scan several times in order to scan the whole storage once is counterintuitive

Reply:
It is temporal solution due to IGNITE-18466. After fix need to change return type to CompositePublisher and the method will be invoked just once. However you are right the require shed lights for the current situation.